tools/k8s_versions/update_k8s_versions.go

```diff
@@ -0,0 +1,142 @@
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("unexpected status code: %d", resp.StatusCode)
+	}
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read response body: %w", err)
+	}
+
+	var kubernetesVersions []KubernetesVersion
+	if err := json.Unmarshal(body, &kubernetesVersions); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal JSON: %w", err)
+	}
+
+	return kubernetesVersions, nil
+}
+
+// yaml parsing does not preserve the original format, mainly blank lines.
+// This function reads the raw yaml file, locate and update the k8s-version block
+func updateK8sVersion(filePath string, kubernetesVersions []KubernetesVersion) error {
+	file, err := os.Open(filePath)
+	if err != nil {
+		return fmt.Errorf("failed to open file: %w", err)
+	}
+	defer file.Close()
+
+	var output []string
+	var k8sVersionBlock []string
+	var inK8sVersionBlock bool
+
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		line := scanner.Text()
+		trimmed := strings.TrimSpace(line)
+
+		// Locate the k8s-version block needed to update
+		if strings.HasPrefix(trimmed, "k8s-version:") || strings.HasPrefix(trimmed, "kubernetes_version:") {
+			inK8sVersionBlock = true
+			output = append(output, line)
+			continue
+		}
+
+		if inK8sVersionBlock {
+			if strings.HasPrefix(trimmed, "-") || strings.HasPrefix(trimmed, "#") {
+				k8sVersionBlock = append(k8sVersionBlock, line)
+				continue
+			} else if trimmed == "" || !strings.HasPrefix(trimmed, "-") {
+				updatedVerBlock, err := updateVersionsBlock(kubernetesVersions, k8sVersionBlock)
+				if err != nil {
+					return err
+				}
+				output = append(output, append(updatedVerBlock, line)...)
+				inK8sVersionBlock = false
+				continue
+			}
+		}
+		output = append(output, line)
+	}
+
+	if err := scanner.Err(); err != nil {
+		return fmt.Errorf("failed to read file: %w", err)
+	}
+
+	if err := os.WriteFile(filePath, []byte(strings.Join(output, "\n")+"\n"), 0644); err != nil {
+		return fmt.Errorf("failed to write updated file: %w", err)
+	}
+	return nil
+}
+
+func updateVersionsBlock(versions []KubernetesVersion, versionBlock []string) ([]string, error) {
+	var output []string
+	now := time.Now()
+	const layout = "2006-01-02"
+
+	for _, version := range versions {
+		eolDate, err := time.Parse(layout, version.EOLDate)
+		if err != nil {
+			return nil, fmt.Errorf("error parsing date: %w", err)
+		}
+
+		for i, line := range versionBlock {
+			trimmed := strings.TrimSpace(line)
+			if strings.HasPrefix(trimmed, fmt.Sprintf("- v%s", version.Cycle)) {
+				if eolDate.After(now) {
+					output = append(output, line)
+				}
+				break
+			} else if i == len(versionBlock)-1 && eolDate.After(now) {
+				// Add the new version at the end of the list
+				output = append(output, fmt.Sprintf("%s- v%s # EOL %s", strings.Repeat(" ", len(line)-len(trimmed)), version.Latest, version.EOLDate))
```

Comment:
We are appending to the list of kind cluster versions we run functional tests against, right? If that's the case we should confirm that kind has an image out for the new version we are adding here or we'll have errors when we try to start in the kind cluster.

Reply:
Good point @jinja2 . I will add some checks and if not available go down to previous minor version

Reply:
PR updated; we now check docker hub if the image exists and decrement the minor minor until we find one